synapseclient/api/configuration_services.py

```diff
@@ -0,0 +1,136 @@
+"""This module is responsible for exposing access to any configuration either through
+file, environment variables, or other means.
+"""
+
+import configparser
+import functools
+from typing import Dict
+
+from synapseclient.core.constants import config_file_constants
+from synapseclient.core.pool_provider import DEFAULT_NUM_THREADS
+
+
+@functools.lru_cache()
+def get_config_file(config_path: str) -> configparser.RawConfigParser:
+    """
+    Retrieves the client configuration information.
+
+    Arguments:
+        config_path:  Path to configuration file on local file system
+
+    Returns:
+        A RawConfigParser populated with properties from the user's configuration file.
+    """
+
+    try:
+        config = configparser.RawConfigParser()
+        config.read(config_path)  # Does not fail if the file does not exist
+        return config
+    except configparser.Error as ex:
+        raise ValueError(f"Error parsing Synapse config file: {config_path}") from ex
+
+
+def get_config_section_dict(
+    section_name: str,
+    config_path: str,
+) -> Dict[str, str]:
+    """
+    Get a profile section in the configuration file with the section name.
+
+    Arguments:
+        section_name: The name of the profile section in the configuration file
+        config_path:  Path to configuration file on local file system
+
+    Returns:
+        A dictionary containing the configuration profile section content
+    """
+    config = get_config_file(config_path)
+    try:
+        return dict(config.items(section_name))
+    except configparser.NoSectionError:
+        # section not present
+        return {}
+
+
+def get_client_authenticated_s3_profile(
+    endpoint: str,
+    bucket: str,
+    config_path: str,
+) -> Dict[str, str]:
+    """
+    Get the authenticated S3 profile from the configuration file.
+
+    Arguments:
+        endpoint: The location of the target service
+        bucket:   AWS S3 bucket name
+        config_path:  Path to configuration file on local file system
+
+    Returns:
+        The authenticated S3 profile
+    """
+
+    config_section = endpoint + "/" + bucket
```

Comment:
Do we want to use `os.path.join(endpoint, bucket)` instead? Manually adding forward slashes like this makes the code a bit more flimsy because you're relying on `endpoint` not already having the forward slash.

Here is an example of why `os.path.join` is more robust:

<img width="305" alt="image" src="https://github.com/Sage-Bionetworks/synapsePythonClient/assets/32107699/61e233d1-b96a-4b81-af31-f05e9757905f">

Reply:
Nit: urllib?

Reply:
> Manually adding forward slashes like this makes the code a bit more flimsy because you're relying on endpoint not already having the forward slash.

I agree - However this relies on the contract that the platform team set in the rest API when we are looking at specific files `fileHandle["endpointUrl"]`

Reply:
Swapped over to urllib to handle this. Verified I could still upload to an external object store I set up before:

![image](https://github.com/Sage-Bionetworks/synapsePythonClient/assets/17128019/42041ce5-b32a-4985-ba38-a6c2cbf43ed5)

`Uploading to endpoint: [https://s3.amazonaws.com] bucket: [sc-237179673806-pp-3ywcz7qgqp4tk-s3bucket-s7wyruiz8llz]: 100%|█|`